// @taiga-ui/experimental/esm2015/directives/avatar-outline/avatar-outline.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TuiAvatarOutlineComponent } from './avatar-outline.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiAvatarOutlineDirective {
    constructor(directiveStyles) {
        this.tuiAvatarOutline = '';
        directiveStyles.addComponent(TuiAvatarOutlineComponent);
    }
    get value() {
        return this.tuiAvatarOutline === ''
            ? 'var(--tui-primary)'
            : this.tuiAvatarOutline;
    }
}
TuiAvatarOutlineDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiAvatarOutlineDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAvatarOutlineDirective, selector: "[tuiAvatarOutline]", inputs: { tuiAvatarOutline: "tuiAvatarOutline" }, host: { properties: { "style.--t-fill": "value", "class._outline": "value" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiAvatarOutline]',
                    host: {
                        '[style.--t-fill]': 'value',
                        '[class._outline]': 'value',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { tuiAvatarOutline: [{
                type: Input
            }] } });
