// @taiga-ui/experimental/bundles/taiga-ui-experimental-components-status.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/components/status', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.components = global["taiga-ui"].experimental.components || {}, global["taiga-ui"].experimental.components.status = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiStatusComponent = /** @class */ (function () {
            function TuiStatusComponent() {
            }
            return TuiStatusComponent;
        }());
        TuiStatusComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStatusComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiStatusComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiStatusComponent, selector: "ng-component", host: { classAttribute: "tui-status" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiStatus]{display:inline-flex;align-items:center;grid-gap:.5rem;gap:.5rem}[tuiStatus]:before{width:.5rem;height:.5rem;border-radius:100%;background:var(--t-status)}[tuiStatus][style*=\"--t-status:\"]:before{content:\"\"}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStatusComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./status.style.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-status',
                            },
                        }]
                }] });

        var TuiStatusDirective = /** @class */ (function () {
            function TuiStatusDirective(directiveStyles) {
                this.tuiStatus = '';
                directiveStyles.addComponent(TuiStatusComponent);
            }
            return TuiStatusDirective;
        }());
        TuiStatusDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStatusDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiStatusDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiStatusDirective, selector: "[tuiStatus]", inputs: { tuiStatus: "tuiStatus" }, host: { attributes: { "tuiStatus": "" }, properties: { "style.--t-status": "tuiStatus || null" } }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStatusDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiStatus]',
                            inputs: ['tuiStatus'],
                            host: {
                                tuiStatus: '',
                                '[style.--t-status]': 'tuiStatus || null',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            } });

        var TuiStatusModule = /** @class */ (function () {
            function TuiStatusModule() {
            }
            return TuiStatusModule;
        }());
        TuiStatusModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStatusModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiStatusModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStatusModule, declarations: [TuiStatusComponent, TuiStatusDirective], exports: [TuiStatusDirective] });
        TuiStatusModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStatusModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiStatusModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiStatusComponent, TuiStatusDirective],
                            exports: [TuiStatusDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiStatusComponent = TuiStatusComponent;
        exports.TuiStatusDirective = TuiStatusDirective;
        exports.TuiStatusModule = TuiStatusModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
