// @taiga-ui/experimental/bundles/taiga-ui-experimental-directives-avatar-outline.umd.js
(function (global, factory) {
        typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk')) :
        typeof define === 'function' && define.amd ? define('@taiga-ui/experimental/directives/avatar-outline', ['exports', '@angular/core', '@taiga-ui/cdk'], factory) :
        (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].experimental = global["taiga-ui"].experimental || {}, global["taiga-ui"].experimental.directives = global["taiga-ui"].experimental.directives || {}, global["taiga-ui"].experimental.directives["avatar-outline"] = {}), global.ng.core, global.i1));
})(this, (function (exports, i0, i1) { 'use strict';

        function _interopNamespace(e) {
                if (e && e.__esModule) return e;
                var n = Object.create(null);
                if (e) {
                        Object.keys(e).forEach(function (k) {
                                if (k !== 'default') {
                                        var d = Object.getOwnPropertyDescriptor(e, k);
                                        Object.defineProperty(n, k, d.get ? d : {
                                                enumerable: true,
                                                get: function () { return e[k]; }
                                        });
                                }
                        });
                }
                n["default"] = e;
                return Object.freeze(n);
        }

        var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
        var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

        var TuiAvatarOutlineComponent = /** @class */ (function () {
            function TuiAvatarOutlineComponent() {
            }
            return TuiAvatarOutlineComponent;
        }());
        TuiAvatarOutlineComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarOutlineComponent, deps: [], target: i0__namespace.ɵɵFactoryTarget.Component });
        TuiAvatarOutlineComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAvatarOutlineComponent, selector: "ng-component", host: { classAttribute: "tui-avatar-outline" }, ngImport: i0__namespace, template: '', isInline: true, styles: ["[tuiAvatarOutline]{--t-outline: .1875rem;--t-gap: .125rem}[tuiAvatarOutline][data-size=xs],[tuiAvatarOutline][data-size=s],[tuiAvatarOutline][data-size=m]{--t-outline: .125rem;--t-gap: .0625rem}[tuiAvatarOutline]._outline{-webkit-mask:radial-gradient(closest-side,#000,#000 calc(100% - var(--t-gap) - var(--t-outline) - .5px),transparent calc(100% - var(--t-gap) - var(--t-outline)),transparent calc(100% - var(--t-outline) - .5px),#000 calc(100% - var(--t-outline)));mask:radial-gradient(closest-side,#000,#000 calc(100% - var(--t-gap) - var(--t-outline) - .5px),transparent calc(100% - var(--t-gap) - var(--t-outline)),transparent calc(100% - var(--t-outline) - .5px),#000 calc(100% - var(--t-outline)))}[tuiAvatarOutline]._outline:after{content:\"\";position:absolute;top:0;left:0;width:100%;height:100%;background:var(--t-fill);-webkit-mask:radial-gradient(closest-side,transparent,transparent calc(100% - var(--t-outline) - .5px),#000 calc(100% - var(--t-outline)));mask:radial-gradient(closest-side,transparent,transparent calc(100% - var(--t-outline) - .5px),#000 calc(100% - var(--t-outline)))}\n"], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush, encapsulation: i0__namespace.ViewEncapsulation.None });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarOutlineComponent, decorators: [{
                    type: i0.Component,
                    args: [{
                            template: '',
                            styleUrls: ['./avatar-outline.styles.less'],
                            encapsulation: i0.ViewEncapsulation.None,
                            changeDetection: i0.ChangeDetectionStrategy.OnPush,
                            host: {
                                class: 'tui-avatar-outline',
                            },
                        }]
                }] });

        var TuiAvatarOutlineDirective = /** @class */ (function () {
            function TuiAvatarOutlineDirective(directiveStyles) {
                this.tuiAvatarOutline = '';
                directiveStyles.addComponent(TuiAvatarOutlineComponent);
            }
            Object.defineProperty(TuiAvatarOutlineDirective.prototype, "value", {
                get: function () {
                    return this.tuiAvatarOutline === ''
                        ? 'var(--tui-primary)'
                        : this.tuiAvatarOutline;
                },
                enumerable: false,
                configurable: true
            });
            return TuiAvatarOutlineDirective;
        }());
        TuiAvatarOutlineDirective.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarOutlineDirective, deps: [{ token: i1.TuiDirectiveStylesService }], target: i0__namespace.ɵɵFactoryTarget.Directive });
        TuiAvatarOutlineDirective.ɵdir = i0__namespace.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAvatarOutlineDirective, selector: "[tuiAvatarOutline]", inputs: { tuiAvatarOutline: "tuiAvatarOutline" }, host: { properties: { "style.--t-fill": "value", "class._outline": "value" } }, ngImport: i0__namespace });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarOutlineDirective, decorators: [{
                    type: i0.Directive,
                    args: [{
                            selector: '[tuiAvatarOutline]',
                            host: {
                                '[style.--t-fill]': 'value',
                                '[class._outline]': 'value',
                            },
                        }]
                }], ctorParameters: function () {
                return [{ type: i1__namespace.TuiDirectiveStylesService, decorators: [{
                                type: i0.Inject,
                                args: [i1.TuiDirectiveStylesService]
                            }] }];
            }, propDecorators: { tuiAvatarOutline: [{
                        type: i0.Input
                    }] } });

        var TuiAvatarOutlineModule = /** @class */ (function () {
            function TuiAvatarOutlineModule() {
            }
            return TuiAvatarOutlineModule;
        }());
        TuiAvatarOutlineModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarOutlineModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
        TuiAvatarOutlineModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarOutlineModule, declarations: [TuiAvatarOutlineComponent, TuiAvatarOutlineDirective], exports: [TuiAvatarOutlineDirective] });
        TuiAvatarOutlineModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarOutlineModule });
        i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiAvatarOutlineModule, decorators: [{
                    type: i0.NgModule,
                    args: [{
                            declarations: [TuiAvatarOutlineComponent, TuiAvatarOutlineDirective],
                            exports: [TuiAvatarOutlineDirective],
                        }]
                }] });

        /**
         * Generated bundle index. Do not edit.
         */

        exports.TuiAvatarOutlineComponent = TuiAvatarOutlineComponent;
        exports.TuiAvatarOutlineDirective = TuiAvatarOutlineDirective;
        exports.TuiAvatarOutlineModule = TuiAvatarOutlineModule;

        Object.defineProperty(exports, '__esModule', { value: true });

}));
