// @taiga-ui/addon-mobile/esm2015/components/mobile-calendar-dialog/mobile-calendar-dialog.module.js
import { NgModule } from '@angular/core';
import { TuiMobileCalendarModule } from '@taiga-ui/addon-mobile/components/mobile-calendar';
import { TuiActiveZoneModule } from '@taiga-ui/cdk';
import { TUI_MOBILE_CALENDAR } from '@taiga-ui/kit';
import { TuiMobileCalendarDialogComponent } from './mobile-calendar-dialog.component';
import { TuiMobileCalendarDropdownComponent } from './mobile-calendar-dropdown.component';
import * as i0 from "@angular/core";
export class TuiMobileCalendarDialogModule {
}
TuiMobileCalendarDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiMobileCalendarDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarDialogModule, declarations: [TuiMobileCalendarDialogComponent, TuiMobileCalendarDropdownComponent], imports: [TuiMobileCalendarModule, TuiActiveZoneModule], exports: [TuiMobileCalendarDialogComponent] });
TuiMobileCalendarDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarDialogModule, providers: [
        {
            provide: TUI_MOBILE_CALENDAR,
            useValue: TuiMobileCalendarDropdownComponent,
        },
    ], imports: [[TuiMobileCalendarModule, TuiActiveZoneModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiMobileCalendarDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [TuiMobileCalendarModule, TuiActiveZoneModule],
                    declarations: [TuiMobileCalendarDialogComponent, TuiMobileCalendarDropdownComponent],
                    providers: [
                        {
                            provide: TUI_MOBILE_CALENDAR,
                            useValue: TuiMobileCalendarDropdownComponent,
                        },
                    ],
                    exports: [TuiMobileCalendarDialogComponent],
                }]
        }] });
