// @taiga-ui/cdk/esm2015/enums/inject-flags.js
/**
 * In Angular v20 InjectFlags is no longer public api
 * This enum is a complete copy to keep things working
 * @deprecated Use `InjectFlags` from `@angular/core` instead!
 * @see https://github.com/angular/angular/blob/12.3.x/packages/compiler/src/core.ts#L261-L280
 */
// eslint-disable-next-line no-restricted-syntax,@typescript-eslint/naming-convention
export var InjectFlags;
(function (InjectFlags) {
    /** Check self and check parent injector if needed */
    InjectFlags[InjectFlags["Default"] = 0] = "Default";
    /**
     * Specifies that an injector should retrieve a dependency from any injector until reaching the
     * host element of the current component. (Only used with Element Injector)
     */
    InjectFlags[InjectFlags["Host"] = 1] = "Host";
    /** Don't ascend to ancestors of the node requesting injection. */
    InjectFlags[InjectFlags["Self"] = 2] = "Self";
    /** Skip the node that is requesting injection. */
    InjectFlags[InjectFlags["SkipSelf"] = 4] = "SkipSelf";
    /** Inject `defaultValue` instead if token not found. */
    InjectFlags[InjectFlags["Optional"] = 8] = "Optional";
})(InjectFlags || (InjectFlags = {}));
