// @taiga-ui/kit/esm2015/abstract/abstract-native-select.js
import { Directive, ElementRef, HostBinding, Inject, Input, TemplateRef, ViewChild, } from '@angular/core';
import { AbstractTuiControl, TuiIdService } from '@taiga-ui/cdk';
import { TUI_TEXTFIELD_HOST, TuiDataListDirective } from '@taiga-ui/core';
import { TUI_ITEMS_HANDLERS } from '@taiga-ui/kit/tokens';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class AbstractTuiNativeSelect {
    constructor(host, control, el, idService, itemsHandlers) {
        this.host = host;
        this.control = control;
        this.el = el;
        this.idService = idService;
        this.itemsHandlers = itemsHandlers;
        this.placeholder = '';
        this.disabledItemHandler = null;
        this.datalist = null;
    }
    get id() {
        return this.el.nativeElement.id || this.idService.generate();
    }
    get emptyOption() {
        return !!this.placeholder && !this.control.value;
    }
}
AbstractTuiNativeSelect.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiNativeSelect, deps: [{ token: TUI_TEXTFIELD_HOST }, { token: AbstractTuiControl }, { token: ElementRef }, { token: TuiIdService }, { token: TUI_ITEMS_HANDLERS }], target: i0.ɵɵFactoryTarget.Directive });
AbstractTuiNativeSelect.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: AbstractTuiNativeSelect, inputs: { placeholder: "placeholder", disabledItemHandler: "disabledItemHandler" }, host: { properties: { "id": "this.id" } }, viewQueries: [{ propertyName: "datalist", first: true, predicate: TuiDataListDirective, descendants: true, read: TemplateRef, static: true }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: AbstractTuiNativeSelect, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_TEXTFIELD_HOST]
                }] }, { type: i1.AbstractTuiControl, decorators: [{
                    type: Inject,
                    args: [AbstractTuiControl]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i1.TuiIdService, decorators: [{
                    type: Inject,
                    args: [TuiIdService]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ITEMS_HANDLERS]
                }] }]; }, propDecorators: { placeholder: [{
                type: Input
            }], disabledItemHandler: [{
                type: Input
            }], datalist: [{
                type: ViewChild,
                args: [TuiDataListDirective, { read: TemplateRef, static: true }]
            }], id: [{
                type: HostBinding,
                args: ['id']
            }] } });
