// @taiga-ui/cdk/bundles/taiga-ui-cdk-enums.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/enums', ['exports'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"].cdk = global["taiga-ui"].cdk || {}, global["taiga-ui"].cdk.enums = {})));
})(this, (function (exports) { 'use strict';

    // TODO: change type in v4.0
    // eslint-disable-next-line no-restricted-syntax
    exports.TuiDayOfWeek = void 0;
    (function (TuiDayOfWeek) {
        TuiDayOfWeek[TuiDayOfWeek["Sunday"] = 0] = "Sunday";
        TuiDayOfWeek[TuiDayOfWeek["Monday"] = 1] = "Monday";
        TuiDayOfWeek[TuiDayOfWeek["Tuesday"] = 2] = "Tuesday";
        TuiDayOfWeek[TuiDayOfWeek["Wednesday"] = 3] = "Wednesday";
        TuiDayOfWeek[TuiDayOfWeek["Thursday"] = 4] = "Thursday";
        TuiDayOfWeek[TuiDayOfWeek["Friday"] = 5] = "Friday";
        TuiDayOfWeek[TuiDayOfWeek["Saturday"] = 6] = "Saturday";
    })(exports.TuiDayOfWeek || (exports.TuiDayOfWeek = {}));

    /**
     * In Angular v20 InjectFlags is no longer public api
     * This enum is a complete copy to keep things working
     * @deprecated Use `InjectFlags` from `@angular/core` instead!
     * @see https://github.com/angular/angular/blob/12.3.x/packages/compiler/src/core.ts#L261-L280
     */
    // eslint-disable-next-line no-restricted-syntax,@typescript-eslint/naming-convention
    exports.InjectFlags = void 0;
    (function (InjectFlags) {
        /** Check self and check parent injector if needed */
        InjectFlags[InjectFlags["Default"] = 0] = "Default";
        /**
         * Specifies that an injector should retrieve a dependency from any injector until reaching the
         * host element of the current component. (Only used with Element Injector)
         */
        InjectFlags[InjectFlags["Host"] = 1] = "Host";
        /** Don't ascend to ancestors of the node requesting injection. */
        InjectFlags[InjectFlags["Self"] = 2] = "Self";
        /** Skip the node that is requesting injection. */
        InjectFlags[InjectFlags["SkipSelf"] = 4] = "SkipSelf";
        /** Inject `defaultValue` instead if token not found. */
        InjectFlags[InjectFlags["Optional"] = 8] = "Optional";
    })(exports.InjectFlags || (exports.InjectFlags = {}));

    // TODO: change type in v4.0
    // eslint-disable-next-line no-restricted-syntax
    exports.TuiMonthNumber = void 0;
    (function (TuiMonthNumber) {
        TuiMonthNumber[TuiMonthNumber["January"] = 0] = "January";
        TuiMonthNumber[TuiMonthNumber["February"] = 1] = "February";
        TuiMonthNumber[TuiMonthNumber["March"] = 2] = "March";
        TuiMonthNumber[TuiMonthNumber["April"] = 3] = "April";
        TuiMonthNumber[TuiMonthNumber["May"] = 4] = "May";
        TuiMonthNumber[TuiMonthNumber["June"] = 5] = "June";
        TuiMonthNumber[TuiMonthNumber["July"] = 6] = "July";
        TuiMonthNumber[TuiMonthNumber["August"] = 7] = "August";
        TuiMonthNumber[TuiMonthNumber["September"] = 8] = "September";
        TuiMonthNumber[TuiMonthNumber["October"] = 9] = "October";
        TuiMonthNumber[TuiMonthNumber["November"] = 10] = "November";
        TuiMonthNumber[TuiMonthNumber["December"] = 11] = "December";
    })(exports.TuiMonthNumber || (exports.TuiMonthNumber = {}));

    /**
     * Generated bundle index. Do not edit.
     */

    Object.defineProperty(exports, '__esModule', { value: true });

}));
