// @taiga-ui/addon-mobile/esm2015/components/sheet-dialog/sheet-dialog.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, HostListener, Inject, ViewChild, ViewChildren, } from '@angular/core';
import { EMPTY_QUERY, tuiPure } from '@taiga-ui/cdk';
import { TUI_ANIMATIONS_DURATION, TUI_CLOSE_WORD, TUI_COMMON_ICONS, tuiSlideInTop, } from '@taiga-ui/core';
import { shouldCall } from '@tinkoff/ng-event-plugins';
import { POLYMORPHEUS_CONTEXT } from '@tinkoff/ng-polymorpheus';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "@taiga-ui/cdk";
import * as i3 from "./sheet-dialog-close.directive";
import * as i4 from "@tinkoff/ng-polymorpheus";
import * as i5 from "rxjs";
// So we re-enter ngZone and trigger change detection
function isCloseable() {
    return this.context.closeable;
}
export class TuiSheetDialogComponent {
    constructor(el, duration, icons, closeWord$, context) {
        this.el = el;
        this.duration = duration;
        this.icons = icons;
        this.closeWord$ = closeWord$;
        this.context = context;
        this.stopsRefs = EMPTY_QUERY;
        this.pointers = 0;
        this.slideInTop = {
            value: '',
            params: {
                start: '100vh',
                duration: this.duration,
            },
        };
        this.stuck$ = new BehaviorSubject(false);
    }
    get closeable() {
        return this.context.closeable;
    }
    get isSmall() {
        var _a;
        return this.sheetTop > (((_a = this.sheet) === null || _a === void 0 ? void 0 : _a.nativeElement.clientHeight) || Infinity);
    }
    onPointerChange(delta) {
        this.pointers = Math.max(this.pointers + delta, 0);
        if (!delta) {
            const stuck = this.el.nativeElement.scrollTop > this.sheetTop;
            this.stuck$.value !== stuck && this.stuck$.next(stuck);
        }
        if (this.context.closeable &&
            !this.pointers &&
            this.el.nativeElement.scrollTop <= 0) {
            this.close();
        }
    }
    close() {
        // TODO: Refactor focus visible on mobile
        this.el.nativeElement.dispatchEvent(new Event('mousedown', { bubbles: true }));
        this.context.$implicit.complete();
    }
    ngAfterViewInit() {
        this.el.nativeElement.scrollTop = [
            ...this.getStops(this.stopsRefs),
            this.sheetTop,
        ][this.context.initial];
    }
    get sheetTop() {
        var _a, _b;
        return (_b = (_a = this.sheet) === null || _a === void 0 ? void 0 : _a.nativeElement.offsetTop) !== null && _b !== void 0 ? _b : Infinity;
    }
    getStops(stops) {
        return stops.map(({ nativeElement }) => nativeElement.offsetTop + nativeElement.clientHeight);
    }
}
TuiSheetDialogComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogComponent, deps: [{ token: ElementRef }, { token: TUI_ANIMATIONS_DURATION }, { token: TUI_COMMON_ICONS }, { token: TUI_CLOSE_WORD }, { token: POLYMORPHEUS_CONTEXT }], target: i0.ɵɵFactoryTarget.Component });
TuiSheetDialogComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetDialogComponent, selector: "tui-sheet-dialog", host: { listeners: { "$.class._stuck": "stuck$", "document:touchstart.passive.silent": "onPointerChange(1)", "document:touchend.silent": "onPointerChange(-1)", "document:touchcancel.silent": "onPointerChange(-1)", "scroll.silent": "onPointerChange(0)" }, properties: { "$.class._stuck": "stuck$", "style.--tui-offset.px": "context.offset", "class._fullscreen": "context.fullscreen === true", "attr.data-appearance": "context.appearance", "@tuiSlideInTop": "this.slideInTop", "class._closeable": "this.closeable" } }, viewQueries: [{ propertyName: "sheet", first: true, predicate: ["sheet"], descendants: true }, { propertyName: "stopsRefs", predicate: ["stops"], descendants: true }], ngImport: i0, template: "<div class=\"t-stops\">\n    <div\n        *ngFor=\"let stop of context.stops\"\n        #stops\n        class=\"t-stop\"\n        [style.margin-top]=\"stop\"\n    ></div>\n</div>\n<div\n    #sheet\n    class=\"t-sheet\"\n    [class.t-sheet_small]=\"isSmall\"\n    (tuiClickOutside)=\"close()\"\n    (tuiSheetDialogClose)=\"close()\"\n>\n    <div\n        *ngIf=\"context.bar\"\n        class=\"t-top\"\n    ></div>\n    <h2\n        *ngIf=\"context.label\"\n        class=\"t-heading\"\n        [id]=\"context.id\"\n    >\n        <ng-container *polymorpheusOutlet=\"context.label as label; context: context\">\n            {{ label }}\n        </ng-container>\n    </h2>\n    <div\n        *ngIf=\"context.content\"\n        class=\"t-content\"\n    >\n        <ng-container *polymorpheusOutlet=\"context.content as text; context: context\">\n            {{ text }}\n        </ng-container>\n    </div>\n</div>\n<div class=\"t-footer\"></div>\n", styles: [":host{position:fixed;top:0;left:0;bottom:0;right:0;right:-1rem;top:calc(var(--tui-offset) + env(safe-area-inset-top));display:flex;flex-direction:column;-webkit-clip-path:inset(0 1rem 0 0 round .75rem .75rem 0 0);clip-path:inset(0 1rem 0 0 round .75rem .75rem 0 0);block-size:calc(100% - var(--tui-offset) - env(safe-area-inset-top));font:var(--tui-font-text-m);overflow-y:scroll;scroll-snap-type:y mandatory;overscroll-behavior:none;box-shadow:0 20rem var(--tui-elevation-01)}:host._closeable{display:block}:host._closeable .t-stops{display:flex}:host._fullscreen{display:block}:host._fullscreen .t-sheet{display:flex;flex-direction:column;min-block-size:calc(100% - 1rem - env(safe-area-inset-bottom))}:host._fullscreen .t-content{display:flex;flex-direction:column;flex-grow:1}.t-stops{display:none;height:100%;scroll-snap-stop:always;scroll-snap-align:start}.t-stop{top:env(safe-area-inset-bottom);scroll-snap-stop:normal;scroll-snap-align:start;height:1rem;width:1rem}.t-sheet{box-shadow:var(--tui-shadow);width:calc(100% - 1rem);border-radius:.75rem .75rem 0 0;padding:0 1rem;margin-top:auto;background:var(--tui-elevation-01);box-sizing:border-box;scroll-snap-stop:always;scroll-snap-align:start}@supports (-moz-appearance: none){.t-sheet_small{scroll-snap-align:end}}.t-top{position:-webkit-sticky;position:sticky;top:0;z-index:1;height:1.5rem;background:var(--tui-elevation-01)}.t-top:after{content:\"\";position:absolute;top:.5rem;left:50%;width:2rem;height:.25rem;transform:translate(-50%);background:var(--tui-base-09);opacity:.3;border-radius:1rem}.t-heading{position:-webkit-sticky;position:sticky;top:1.5rem;z-index:1;display:flex;margin:0;padding-bottom:1rem;font:var(--tui-font-heading-6);background:var(--tui-elevation-01)}.t-heading:last-child{margin-bottom:-.75rem}.t-heading:after{content:\"\";position:absolute;top:100%;left:0;right:0;height:1px;background:var(--tui-base-03);opacity:0}:host._stuck .t-heading:after{opacity:1}.t-close{right:-.25rem;flex-shrink:0;margin-left:auto}.t-content{position:relative;isolation:isolate;border-radius:inherit}.t-content:nth-child(3){margin-top:1rem}.t-footer{height:calc(1rem + env(safe-area-inset-bottom));scroll-snap-stop:always;scroll-snap-align:end;background:var(--tui-elevation-01)}\n"], directives: [{ type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { type: i2.TuiClickOutsideDirective, selector: "[tuiClickOutside]", outputs: ["tuiClickOutside"] }, { type: i3.TuiSheetDialogCloseDirective, selector: "[tuiSheetDialogClose]", outputs: ["tuiSheetDialogClose"] }, { type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4.PolymorpheusOutletDirective, selector: "[polymorpheusOutlet]", inputs: ["polymorpheusOutlet", "polymorpheusOutletContext"] }], animations: [tuiSlideInTop], changeDetection: i0.ChangeDetectionStrategy.OnPush });
__decorate([
    shouldCall(isCloseable)
], TuiSheetDialogComponent.prototype, "close", null);
__decorate([
    tuiPure
], TuiSheetDialogComponent.prototype, "getStops", null);
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-sheet-dialog',
                    templateUrl: './sheet-dialog.template.html',
                    styleUrls: ['./sheet-dialog.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [tuiSlideInTop],
                    host: {
                        '[$.class._stuck]': 'stuck$',
                        '($.class._stuck)': 'stuck$',
                        '[style.--tui-offset.px]': 'context.offset',
                        '[class._fullscreen]': 'context.fullscreen === true',
                        '[attr.data-appearance]': 'context.appearance',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ANIMATIONS_DURATION]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_COMMON_ICONS]
                }] }, { type: i5.Observable, decorators: [{
                    type: Inject,
                    args: [TUI_CLOSE_WORD]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [POLYMORPHEUS_CONTEXT]
                }] }]; }, propDecorators: { sheet: [{
                type: ViewChild,
                args: ['sheet']
            }], stopsRefs: [{
                type: ViewChildren,
                args: ['stops']
            }], slideInTop: [{
                type: HostBinding,
                args: ['@tuiSlideInTop']
            }], closeable: [{
                type: HostBinding,
                args: ['class._closeable']
            }], onPointerChange: [{
                type: HostListener,
                args: ['document:touchstart.passive.silent', ['1']]
            }, {
                type: HostListener,
                args: ['document:touchend.silent', ['-1']]
            }, {
                type: HostListener,
                args: ['document:touchcancel.silent', ['-1']]
            }, {
                type: HostListener,
                args: ['scroll.silent', ['0']]
            }], close: [], getStops: [] } });
