// @taiga-ui/addon-mobile/esm2015/components/sheet-dialog/sheet-dialog-close.directive.js
import { Directive, ElementRef, Inject, NgZone, Output } from '@angular/core';
import { tuiTypedFromEvent, tuiZonefree } from '@taiga-ui/cdk';
import * as i0 from "@angular/core";
export const TUI_SHEET_DIALOG_CLOSE = 'tui-sheet-dialog-close';
export class TuiSheetDialogCloseDirective {
    constructor(zone, el) {
        this.zone = zone;
        this.el = el;
        this.tuiSheetDialogClose = tuiTypedFromEvent(this.el.nativeElement, TUI_SHEET_DIALOG_CLOSE).pipe(tuiZonefree(this.zone));
    }
}
TuiSheetDialogCloseDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogCloseDirective, deps: [{ token: NgZone }, { token: ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
TuiSheetDialogCloseDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiSheetDialogCloseDirective, selector: "[tuiSheetDialogClose]", outputs: { tuiSheetDialogClose: "tuiSheetDialogClose" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiSheetDialogCloseDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiSheetDialogClose]',
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { tuiSheetDialogClose: [{
                type: Output
            }] } });
