// @taiga-ui/experimental/esm2015/directives/cell/cell.directive.js
import { Directive, Inject, Input } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { tuiAvatarOptionsProvider, tuiButtonOptionsProvider, } from '@taiga-ui/experimental/components';
import { TuiCellComponent } from './cell.component';
import { TUI_CELL_OPTIONS } from './cell.options';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiCellDirective {
    constructor(directiveStyles, options) {
        this.options = options;
        this.size = this.options.size;
        this.height = this.options.height;
        directiveStyles.addComponent(TuiCellComponent);
    }
}
TuiCellDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCellDirective, deps: [{ token: TuiDirectiveStylesService }, { token: TUI_CELL_OPTIONS }], target: i0.ɵɵFactoryTarget.Directive });
TuiCellDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiCellDirective, selector: "[tuiCell]:not(ng-template)", inputs: { size: ["tuiCell", "size"], height: ["tuiCellHeight", "height"] }, host: { attributes: { "tuiCell": "" }, properties: { "attr.data-size": "size || options.size", "attr.data-height": "height" } }, providers: [
        tuiAvatarOptionsProvider({ size: 's' }),
        tuiButtonOptionsProvider({ size: 's' }),
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiCellDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiCell]:not(ng-template)',
                    providers: [
                        tuiAvatarOptionsProvider({ size: 's' }),
                        tuiButtonOptionsProvider({ size: 's' }),
                    ],
                    host: {
                        tuiCell: '',
                        '[attr.data-size]': 'size || options.size',
                        '[attr.data-height]': 'height',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_CELL_OPTIONS]
                }] }]; }, propDecorators: { size: [{
                type: Input,
                args: ['tuiCell']
            }], height: [{
                type: Input,
                args: ['tuiCellHeight']
            }] } });
