// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-components-status.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';

class TuiStatusComponent {
}
TuiStatusComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiStatusComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiStatusComponent, selector: "ng-component", host: { classAttribute: "tui-status" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiStatus]{display:inline-flex;align-items:center;grid-gap:.5rem;gap:.5rem}[tuiStatus]:before{width:.5rem;height:.5rem;border-radius:100%;background:var(--t-status)}[tuiStatus][style*=\"--t-status:\"]:before{content:\"\"}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./status.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-status',
                    },
                }]
        }] });

class TuiStatusDirective {
    constructor(directiveStyles) {
        this.tuiStatus = '';
        directiveStyles.addComponent(TuiStatusComponent);
    }
}
TuiStatusDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiStatusDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiStatusDirective, selector: "[tuiStatus]", inputs: { tuiStatus: "tuiStatus" }, host: { attributes: { "tuiStatus": "" }, properties: { "style.--t-status": "tuiStatus || null" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiStatus]',
                    inputs: ['tuiStatus'],
                    host: {
                        tuiStatus: '',
                        '[style.--t-status]': 'tuiStatus || null',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; } });

class TuiStatusModule {
}
TuiStatusModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiStatusModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusModule, declarations: [TuiStatusComponent, TuiStatusDirective], exports: [TuiStatusDirective] });
TuiStatusModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiStatusComponent, TuiStatusDirective],
                    exports: [TuiStatusDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiStatusComponent, TuiStatusDirective, TuiStatusModule };
