// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-components-action-bar.js
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Inject, Input, Directive, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/core';
import { TUI_ANIMATION_OPTIONS, tuiFadeIn, tuiSlideInTop, TuiExpandModule } from '@taiga-ui/core';
import { TuiPortalDirective } from '@taiga-ui/cdk';
import { CommonModule } from '@angular/common';

class TuiActionBarComponent {
    constructor(animation) {
        this.animation = animation;
        this.expanded = false;
    }
}
TuiActionBarComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarComponent, deps: [{ token: TUI_ANIMATION_OPTIONS }], target: i0.ɵɵFactoryTarget.Component });
TuiActionBarComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiActionBarComponent, selector: "tui-action-bar", inputs: { expanded: "expanded" }, host: { attributes: { "tuiTheme": "night" }, properties: { "@tuiFadeIn": "animation", "@tuiSlideInTop": "animation" } }, ngImport: i0, template: "<tui-expand [expanded]=\"expanded\">\n    <ng-content select=\"tui-data-list,[tuiMenu]\"></ng-content>\n</tui-expand>\n\n<div class=\"t-content\">\n    <ng-content></ng-content>\n\n    <div class=\"t-actions\">\n        <ng-content select=\"a,button,[tuiAction]\"></ng-content>\n    </div>\n</div>\n", styles: [":host{position:fixed;left:1rem;right:1rem;bottom:1rem;display:flex;box-sizing:border-box;min-height:4rem;border-radius:var(--tui-radius-xl);background:rgba(0,0,0,.75);color:var(--tui-text-01);-webkit-backdrop-filter:blur(2rem);backdrop-filter:blur(2rem);flex-direction:column;justify-content:center;padding:.625rem}:host ::ng-deep tui-data-list[data-size]{padding:.25rem 0 .5rem}:host[tuiTheme=night]{--tui-text-01: var(--tui-text-01-night);--tui-text-02: var(--tui-text-02-night);--tui-text-03: var(--tui-text-03-night);--tui-base-03: var(--tui-neutral-bg-night)}.t-content{display:flex;justify-content:space-between;align-items:center;padding-left:.625rem}.t-actions{display:flex;grid-gap:.5rem;gap:.5rem}:host-context(tui-root._mobile) .t-actions{grid-gap:.375rem;gap:.375rem}\n"], components: [{ type: i1.TuiExpandComponent, selector: "tui-expand", inputs: ["async", "expanded"] }], animations: [tuiFadeIn, tuiSlideInTop], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-action-bar',
                    templateUrl: './action-bar.template.html',
                    styleUrls: ['./action-bar.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    animations: [tuiFadeIn, tuiSlideInTop],
                    host: {
                        tuiTheme: 'night',
                        '[@tuiFadeIn]': 'animation',
                        '[@tuiSlideInTop]': 'animation',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ANIMATION_OPTIONS]
                }] }]; }, propDecorators: { expanded: [{
                type: Input
            }] } });

class TuiActionBarDirective extends TuiPortalDirective {
}
TuiActionBarDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiActionBarDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiActionBarDirective, selector: "ng-template[tuiActionBar]", inputs: { tuiPortal: ["tuiActionBar", "tuiPortal"] }, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[tuiActionBar]',
                    inputs: ['tuiPortal: tuiActionBar'],
                }]
        }] });

class TuiActionBarModule {
}
TuiActionBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiActionBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarModule, declarations: [TuiActionBarDirective, TuiActionBarComponent], imports: [CommonModule, TuiExpandModule], exports: [TuiActionBarDirective, TuiActionBarComponent] });
TuiActionBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarModule, imports: [[CommonModule, TuiExpandModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiExpandModule],
                    declarations: [TuiActionBarDirective, TuiActionBarComponent],
                    exports: [TuiActionBarDirective, TuiActionBarComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiActionBarComponent, TuiActionBarDirective, TuiActionBarModule };
