// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-components-thumbnail-card.js
import * as i0 from '@angular/core';
import { Component, ChangeDetectionStrategy, Inject, Input, NgModule } from '@angular/core';
import { TUI_ICON_RESOLVER } from '@taiga-ui/experimental/tokens';
import { TUI_PAYMENT_SYSTEM_ICONS } from '@taiga-ui/addon-commerce';
import { tuiCreateToken } from '@taiga-ui/cdk';
import * as i1 from '@taiga-ui/experimental/components/icon';
import { TuiIconModule } from '@taiga-ui/experimental/components/icon';
import * as i2 from '@angular/common';
import { CommonModule } from '@angular/common';

const TUI_THUMBNAIL_CARD_OPTIONS = tuiCreateToken({
    icons: TUI_PAYMENT_SYSTEM_ICONS,
    size: 'm',
    monoHandler: ps => ps === 'mir' || ps === 'visa' || ps === 'electron',
});

class TuiThumbnailCardComponent {
    constructor(resolver, options) {
        this.resolver = resolver;
        this.options = options;
        this.size = this.options.size;
        this.paymentSystem = null;
        this.iconStart = '';
        this.iconEnd = '';
        this.monoHandler = this.options.monoHandler;
    }
}
TuiThumbnailCardComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardComponent, deps: [{ token: TUI_ICON_RESOLVER }, { token: TUI_THUMBNAIL_CARD_OPTIONS }], target: i0.ɵɵFactoryTarget.Component });
TuiThumbnailCardComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiThumbnailCardComponent, selector: "tui-thumbnail-card", inputs: { size: "size", paymentSystem: "paymentSystem", iconStart: ["iconLeft", "iconStart"], iconEnd: ["iconRight", "iconEnd"], monoHandler: "monoHandler" }, host: { properties: { "attr.data-size": "size" } }, ngImport: i0, template: "<tui-icon\n    *ngIf=\"iconStart\"\n    class=\"t-icon t-left\"\n    [icon]=\"iconStart\"\n></tui-icon>\n\n<div class=\"t-left t-icon\">\n    <ng-content select=\"img\"></ng-content>\n</div>\n\n<tui-icon\n    *ngIf=\"iconEnd\"\n    class=\"t-right t-icon\"\n    [icon]=\"iconEnd\"\n></tui-icon>\n\n<div class=\"t-content\">\n    <ng-content></ng-content>\n</div>\n\n<ng-container *ngIf=\"paymentSystem\">\n    <tui-icon\n        *ngIf=\"monoHandler(paymentSystem); else colored\"\n        class=\"t-logo\"\n        [icon]=\"options.icons[paymentSystem]\"\n    ></tui-icon>\n    <ng-template #colored>\n        <img\n            class=\"t-logo\"\n            [alt]=\"paymentSystem\"\n            [src]=\"resolver(options.icons[paymentSystem]).replace('Outline', '')\"\n        />\n    </ng-template>\n</ng-container>\n", styles: [":host{--t-logo-height: .8125rem;position:relative;display:inline-block;flex-shrink:0;background:#000;color:#fff;transform-style:preserve-3d;overflow:hidden;box-shadow:inset 0 0 0 1px var(--tui-background-neutral-1)}:host:before{position:absolute;top:0;left:0;width:100%;height:100%;content:\"\";border-radius:inherit}:host:after{position:absolute;top:0;left:0;width:100%;height:100%;content:\"\";-webkit-backface-visibility:hidden;backface-visibility:hidden;transform:rotateY(180deg) translateZ(1px);color:#b0b0b0;border-radius:inherit;background:linear-gradient(to bottom,currentColor 0 20%,#959595 20% 40%,currentColor 40%)}:host[data-size=xs],:host[data-size=s]{font-size:0;border-radius:.25rem}:host[data-size=xs]{inline-size:2rem;block-size:1.375rem}:host[data-size=s]{inline-size:2.5rem;block-size:1.625rem}:host[data-size=m]{inline-size:3rem;block-size:2rem;border-radius:.3125rem;font:normal .5625rem / var(--t-logo-height) var(--tui-font-text)}:host[data-size=l]{inline-size:3.375rem;block-size:2.25rem;border-radius:.375rem;font:normal .6875rem / var(--t-logo-height) var(--tui-font-text)}.t-left{position:absolute;top:0;left:0;display:flex;transform-origin:top left}:host[data-size=xs] .t-left,:host[data-size=s] .t-left{transform:translate(.125rem,.125rem) scale(.625)}:host[data-size=m] .t-left{transform:translate(.1875rem,.1875rem) scale(.875)}:host[data-size=l] .t-left{transform:translate(.1875rem,.1875rem)}.t-right{position:absolute;top:.1875rem;right:.1875rem}:host[data-size=xs] .t-right,:host[data-size=s] .t-right{display:none}:host[data-size=m] .t-right{inline-size:.875rem;block-size:.875rem}:host[data-size=m] .t-right:after{transform:scale(.5)}:host[data-size=l] .t-right{inline-size:1rem;block-size:1rem}:host[data-size=l] .t-right:after{color:#fff;transform:scale(.625)}.t-right:before{content:\"\";position:absolute;top:0;left:0;width:100%;height:100%;border-radius:100%;background:currentColor;opacity:.35}.t-logo{position:absolute;right:.25rem;bottom:0;block-size:var(--t-logo-height);transform-origin:right}:host[data-size=xs] .t-logo{right:.125rem}:host[data-size=m] .t-logo,:host[data-size=l] .t-logo{bottom:.0625rem}.t-logo:after{-webkit-mask-position:right;mask-position:right}.t-content{position:absolute}:host[data-size=m] .t-content{left:.1875rem;bottom:.0625rem}:host[data-size=l] .t-content{left:.25rem;bottom:.0625rem}.t-icon{inline-size:1rem;block-size:1rem}\n"], components: [{ type: i1.TuiIconComponent, selector: "tui-icon", inputs: ["icon", "background"] }], directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-thumbnail-card',
                    templateUrl: './thumbnail-card.template.html',
                    styleUrls: ['./thumbnail-card.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[attr.data-size]': 'size',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_ICON_RESOLVER]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_THUMBNAIL_CARD_OPTIONS]
                }] }]; }, propDecorators: { size: [{
                type: Input
            }], paymentSystem: [{
                type: Input
            }], iconStart: [{
                type: Input,
                args: ['iconLeft']
            }], iconEnd: [{
                type: Input,
                args: ['iconRight']
            }], monoHandler: [{
                type: Input
            }] } });

class TuiThumbnailCardModule {
}
TuiThumbnailCardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiThumbnailCardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, declarations: [TuiThumbnailCardComponent], imports: [CommonModule, TuiIconModule], exports: [TuiThumbnailCardComponent] });
TuiThumbnailCardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, imports: [[CommonModule, TuiIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiThumbnailCardModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiIconModule],
                    declarations: [TuiThumbnailCardComponent],
                    exports: [TuiThumbnailCardComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TUI_THUMBNAIL_CARD_OPTIONS, TuiThumbnailCardComponent, TuiThumbnailCardModule };
