// @taiga-ui/experimental/fesm2015/taiga-ui-experimental-directives-avatar-outline.js
import * as i0 from '@angular/core';
import { Component, ViewEncapsulation, ChangeDetectionStrategy, Directive, Inject, Input, NgModule } from '@angular/core';
import * as i1 from '@taiga-ui/cdk';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';

class TuiAvatarOutlineComponent {
}
TuiAvatarOutlineComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiAvatarOutlineComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiAvatarOutlineComponent, selector: "ng-component", host: { classAttribute: "tui-avatar-outline" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiAvatarOutline]{--t-outline: .1875rem;--t-gap: .125rem}[tuiAvatarOutline][data-size=xs],[tuiAvatarOutline][data-size=s],[tuiAvatarOutline][data-size=m]{--t-outline: .125rem;--t-gap: .0625rem}[tuiAvatarOutline]._outline{-webkit-mask:radial-gradient(closest-side,#000,#000 calc(100% - var(--t-gap) - var(--t-outline) - .5px),transparent calc(100% - var(--t-gap) - var(--t-outline)),transparent calc(100% - var(--t-outline) - .5px),#000 calc(100% - var(--t-outline)));mask:radial-gradient(closest-side,#000,#000 calc(100% - var(--t-gap) - var(--t-outline) - .5px),transparent calc(100% - var(--t-gap) - var(--t-outline)),transparent calc(100% - var(--t-outline) - .5px),#000 calc(100% - var(--t-outline)))}[tuiAvatarOutline]._outline:after{content:\"\";position:absolute;top:0;left:0;width:100%;height:100%;background:var(--t-fill);-webkit-mask:radial-gradient(closest-side,transparent,transparent calc(100% - var(--t-outline) - .5px),#000 calc(100% - var(--t-outline)));mask:radial-gradient(closest-side,transparent,transparent calc(100% - var(--t-outline) - .5px),#000 calc(100% - var(--t-outline)))}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./avatar-outline.styles.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-avatar-outline',
                    },
                }]
        }] });

class TuiAvatarOutlineDirective {
    constructor(directiveStyles) {
        this.tuiAvatarOutline = '';
        directiveStyles.addComponent(TuiAvatarOutlineComponent);
    }
    get value() {
        return this.tuiAvatarOutline === ''
            ? 'var(--tui-primary)'
            : this.tuiAvatarOutline;
    }
}
TuiAvatarOutlineDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiAvatarOutlineDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiAvatarOutlineDirective, selector: "[tuiAvatarOutline]", inputs: { tuiAvatarOutline: "tuiAvatarOutline" }, host: { properties: { "style.--t-fill": "value", "class._outline": "value" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiAvatarOutline]',
                    host: {
                        '[style.--t-fill]': 'value',
                        '[class._outline]': 'value',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; }, propDecorators: { tuiAvatarOutline: [{
                type: Input
            }] } });

class TuiAvatarOutlineModule {
}
TuiAvatarOutlineModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiAvatarOutlineModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineModule, declarations: [TuiAvatarOutlineComponent, TuiAvatarOutlineDirective], exports: [TuiAvatarOutlineDirective] });
TuiAvatarOutlineModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiAvatarOutlineModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiAvatarOutlineComponent, TuiAvatarOutlineDirective],
                    exports: [TuiAvatarOutlineDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiAvatarOutlineComponent, TuiAvatarOutlineDirective, TuiAvatarOutlineModule };
