// @taiga-ui/addon-mobile/esm2015/directives/dropdown-mobile/dropdown-mobile.directive.js
import { Directive, HostListener, Inject, inject, Input } from '@angular/core';
import { InjectFlags, TUI_IS_MOBILE, tuiIsHTMLElement } from '@taiga-ui/cdk';
import { TUI_DROPDOWN_COMPONENT } from '@taiga-ui/core';
import { TuiDropdownMobileComponent } from './dropdown-mobile.component';
import * as i0 from "@angular/core";
export class TuiDropdownMobileDirective {
    constructor(isMobile) {
        this.isMobile = isMobile;
        this.tuiDropdownMobile = '';
    }
    onMouseDown(event) {
        if (!this.isMobile ||
            !tuiIsHTMLElement(event.target) ||
            !event.target.matches('input,textarea')) {
            return;
        }
        event.preventDefault();
        event.target.focus({ preventScroll: true });
    }
}
TuiDropdownMobileDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownMobileDirective, deps: [{ token: TUI_IS_MOBILE }], target: i0.ɵɵFactoryTarget.Directive });
TuiDropdownMobileDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDropdownMobileDirective, selector: "[tuiDropdownMobile]", inputs: { tuiDropdownMobile: "tuiDropdownMobile" }, host: { listeners: { "mousedown": "onMouseDown($event)" }, properties: { "style.visibility": "\"visible\"" } }, providers: [
        {
            provide: TUI_DROPDOWN_COMPONENT,
            useFactory: () => inject(TUI_IS_MOBILE)
                ? TuiDropdownMobileComponent
                : inject(TUI_DROPDOWN_COMPONENT, InjectFlags.SkipSelf),
        },
    ], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownMobileDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiDropdownMobile]',
                    providers: [
                        {
                            provide: TUI_DROPDOWN_COMPONENT,
                            useFactory: () => inject(TUI_IS_MOBILE)
                                ? TuiDropdownMobileComponent
                                : inject(TUI_DROPDOWN_COMPONENT, InjectFlags.SkipSelf),
                        },
                    ],
                    host: {
                        '[style.visibility]': '"visible"',
                    },
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_MOBILE]
                }] }]; }, propDecorators: { tuiDropdownMobile: [{
                type: Input
            }], onMouseDown: [{
                type: HostListener,
                args: ['mousedown', ['$event']]
            }] } });
