// @taiga-ui/addon-mobile/esm2015/directives/dropdown-mobile/dropdown-mobile.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { IntersectionObserverModule } from '@ng-web-apis/intersection-observer';
import { TuiSwipeModule } from '@taiga-ui/cdk';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';
import { TuiDropdownMobileComponent } from './dropdown-mobile.component';
import { TuiDropdownMobileDirective } from './dropdown-mobile.directive';
import * as i0 from "@angular/core";
export class TuiDropdownMobileModule {
}
TuiDropdownMobileModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownMobileModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiDropdownMobileModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownMobileModule, declarations: [TuiDropdownMobileComponent, TuiDropdownMobileDirective], imports: [PolymorpheusModule,
        CommonModule,
        TuiSwipeModule,
        IntersectionObserverModule], exports: [TuiDropdownMobileDirective] });
TuiDropdownMobileModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownMobileModule, imports: [[
            PolymorpheusModule,
            CommonModule,
            TuiSwipeModule,
            IntersectionObserverModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownMobileModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        PolymorpheusModule,
                        CommonModule,
                        TuiSwipeModule,
                        IntersectionObserverModule,
                    ],
                    declarations: [TuiDropdownMobileComponent, TuiDropdownMobileDirective],
                    exports: [TuiDropdownMobileDirective],
                }]
        }] });
