// @taiga-ui/experimental/esm2015/components/navigation/hint-aside.directive.js
import { Directive, ElementRef, Inject } from '@angular/core';
import { TuiHintDirective, tuiHintOptionsProvider } from '@taiga-ui/core';
import { TuiAsideComponent } from './aside.component';
import * as i0 from "@angular/core";
import * as i1 from "./aside.component";
import * as i2 from "@taiga-ui/core";
export class TuiHintAsideDirective {
    constructor(el, aside, hint) {
        this.el = el;
        this.aside = aside;
        this.hint = hint;
    }
    ngDoCheck() {
        var _a;
        this.hint.tuiHint = this.aside.tuiNavigationAside
            ? ''
            : (_a = this.el.nativeElement.textContent) === null || _a === void 0 ? void 0 : _a.trim();
    }
}
TuiHintAsideDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintAsideDirective, deps: [{ token: ElementRef }, { token: TuiAsideComponent }, { token: TuiHintDirective }], target: i0.ɵɵFactoryTarget.Directive });
TuiHintAsideDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiHintAsideDirective, selector: "[tuiHint][tuiHintAside]", providers: [tuiHintOptionsProvider({ direction: 'right' })], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiHintAsideDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiHint][tuiHintAside]',
                    providers: [tuiHintOptionsProvider({ direction: 'right' })],
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i1.TuiAsideComponent, decorators: [{
                    type: Inject,
                    args: [TuiAsideComponent]
                }] }, { type: i2.TuiHintDirective, decorators: [{
                    type: Inject,
                    args: [TuiHintDirective]
                }] }]; } });
