// @taiga-ui/experimental/esm2015/components/navigation/navigation.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiScrollbarModule } from '@taiga-ui/core';
import { TuiFadeModule } from '@taiga-ui/experimental/directives/fade';
import { TuiAsideComponent } from './aside.component';
import { TuiHeaderComponent } from './header.component';
import { TuiHintAsideDirective } from './hint-aside.directive';
import { TuiLogoComponent } from './logo.component';
import { TuiMainComponent } from './main.component';
import { TuiNavComponent } from './nav.component';
import * as i0 from "@angular/core";
export class TuiNavigationModule {
}
TuiNavigationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNavigationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiNavigationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNavigationModule, declarations: [TuiHeaderComponent,
        TuiHintAsideDirective,
        TuiLogoComponent,
        TuiMainComponent,
        TuiAsideComponent,
        TuiNavComponent], imports: [CommonModule, TuiScrollbarModule, TuiFadeModule], exports: [TuiHeaderComponent,
        TuiHintAsideDirective,
        TuiLogoComponent,
        TuiMainComponent,
        TuiAsideComponent,
        TuiNavComponent] });
TuiNavigationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNavigationModule, imports: [[CommonModule, TuiScrollbarModule, TuiFadeModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiNavigationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiScrollbarModule, TuiFadeModule],
                    declarations: [
                        TuiHeaderComponent,
                        TuiHintAsideDirective,
                        TuiLogoComponent,
                        TuiMainComponent,
                        TuiAsideComponent,
                        TuiNavComponent,
                    ],
                    exports: [
                        TuiHeaderComponent,
                        TuiHintAsideDirective,
                        TuiLogoComponent,
                        TuiMainComponent,
                        TuiAsideComponent,
                        TuiNavComponent,
                    ],
                }]
        }] });
