// @tinkoff/ng-event-plugins/esm2015/plugins/global.plugin.js
import { DOCUMENT } from '@angular/common';
import { inject, Injectable, InjectionToken } from '@angular/core';
import { AbstractEventPlugin } from './abstract.plugin';
import * as i0 from "@angular/core";
export const GLOBAL_HANDLER = new InjectionToken('[GLOBAL_HANDLER]: Global event target handler', {
    factory: () => {
        const document = inject(DOCUMENT);
        return name => name === 'body'
            ? document.body
            : document.defaultView[name] ||
                document.createElement('div');
    },
});
export class GlobalEventPlugin extends AbstractEventPlugin {
    constructor() {
        super(...arguments);
        this.handler = inject(GLOBAL_HANDLER);
        this.modifier = '>';
    }
    addEventListener(_, event, handler) {
        return this.manager.addEventListener(this.handler(event.split('>')[0]), event.split('>')[1], handler);
    }
}
/** @nocollapse */ GlobalEventPlugin.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: GlobalEventPlugin, deps: null, target: i0.ɵɵFactoryTarget.Injectable });
/** @nocollapse */ GlobalEventPlugin.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: GlobalEventPlugin });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: GlobalEventPlugin, decorators: [{
            type: Injectable
        }] });
