// @taiga-ui/experimental/esm2015/components/action-bar/action-bar.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiExpandModule } from '@taiga-ui/core';
import { TuiActionBarComponent } from './action-bar.component';
import { TuiActionBarDirective } from './action-bar.directive';
import * as i0 from "@angular/core";
export class TuiActionBarModule {
}
TuiActionBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiActionBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarModule, declarations: [TuiActionBarDirective, TuiActionBarComponent], imports: [CommonModule, TuiExpandModule], exports: [TuiActionBarDirective, TuiActionBarComponent] });
TuiActionBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarModule, imports: [[CommonModule, TuiExpandModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiActionBarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, TuiExpandModule],
                    declarations: [TuiActionBarDirective, TuiActionBarComponent],
                    exports: [TuiActionBarDirective, TuiActionBarComponent],
                }]
        }] });
