// @taiga-ui/core/esm2015/components/hosted-dropdown/dropdown-open-monitor.directive.js
import { Directive, Inject, Input, Optional, Self } from '@angular/core';
import { TuiDestroyService } from '@taiga-ui/cdk';
import { TuiDropdownDirective, TuiDropdownOpenDirective } from '@taiga-ui/core/directives';
import { takeUntil } from 'rxjs/operators';
import { TUI_HOSTED_DROPDOWN_COMPONENT } from './hosted-dropdown.token';
import * as i0 from "@angular/core";
import * as i1 from "rxjs";
import * as i2 from "@taiga-ui/core/directives";
export class TuiDropdownOpenMonitorDirective {
    constructor(destroy$, hosted, dropdown, open) {
        this.open = open;
        if ((open === null || open === void 0 ? void 0 : open.dropdown) === dropdown) {
            open.tuiDropdownOpenChange.pipe(takeUntil(destroy$)).subscribe(value => {
                var _a;
                if (value) {
                    (_a = hosted.nativeFocusableElement) === null || _a === void 0 ? void 0 : _a.focus();
                }
                hosted.updateOpen(value);
            });
        }
    }
    set tuiDropdownOpenMonitor(open) {
        var _a;
        (_a = this.open) === null || _a === void 0 ? void 0 : _a.update(open);
    }
}
TuiDropdownOpenMonitorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownOpenMonitorDirective, deps: [{ token: TuiDestroyService, self: true }, { token: TUI_HOSTED_DROPDOWN_COMPONENT }, { token: TuiDropdownDirective, self: true }, { token: TuiDropdownOpenDirective, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
TuiDropdownOpenMonitorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiDropdownOpenMonitorDirective, selector: "[tuiDropdownOpenMonitor]", inputs: { tuiDropdownOpenMonitor: "tuiDropdownOpenMonitor" }, providers: [TuiDestroyService], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiDropdownOpenMonitorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiDropdownOpenMonitor]',
                    providers: [TuiDestroyService],
                }]
        }], ctorParameters: function () { return [{ type: i1.Observable, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDestroyService]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_HOSTED_DROPDOWN_COMPONENT]
                }] }, { type: i2.TuiDropdownDirective, decorators: [{
                    type: Self
                }, {
                    type: Inject,
                    args: [TuiDropdownDirective]
                }] }, { type: i2.TuiDropdownOpenDirective, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [TuiDropdownOpenDirective]
                }] }]; }, propDecorators: { tuiDropdownOpenMonitor: [{
                type: Input
            }] } });
