// @taiga-ui/experimental/esm2015/components/status/status.directive.js
import { Directive, Inject } from '@angular/core';
import { TuiDirectiveStylesService } from '@taiga-ui/cdk';
import { TuiStatusComponent } from './status.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk";
export class TuiStatusDirective {
    constructor(directiveStyles) {
        this.tuiStatus = '';
        directiveStyles.addComponent(TuiStatusComponent);
    }
}
TuiStatusDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusDirective, deps: [{ token: TuiDirectiveStylesService }], target: i0.ɵɵFactoryTarget.Directive });
TuiStatusDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiStatusDirective, selector: "[tuiStatus]", inputs: { tuiStatus: "tuiStatus" }, host: { attributes: { "tuiStatus": "" }, properties: { "style.--t-status": "tuiStatus || null" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[tuiStatus]',
                    inputs: ['tuiStatus'],
                    host: {
                        tuiStatus: '',
                        '[style.--t-status]': 'tuiStatus || null',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i1.TuiDirectiveStylesService, decorators: [{
                    type: Inject,
                    args: [TuiDirectiveStylesService]
                }] }]; } });
