// @taiga-ui/experimental/esm2015/components/status/status.component.js
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
import * as i0 from "@angular/core";
export class TuiStatusComponent {
}
TuiStatusComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TuiStatusComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiStatusComponent, selector: "ng-component", host: { classAttribute: "tui-status" }, ngImport: i0, template: '', isInline: true, styles: ["[tuiStatus]{display:inline-flex;align-items:center;grid-gap:.5rem;gap:.5rem}[tuiStatus]:before{width:.5rem;height:.5rem;border-radius:100%;background:var(--t-status)}[tuiStatus][style*=\"--t-status:\"]:before{content:\"\"}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiStatusComponent, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    styleUrls: ['./status.style.less'],
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'tui-status',
                    },
                }]
        }] });
