// @taiga-ui/addon-mobile/esm2015/directives/responsive-dialog/responsive-dialog.service.js
import { Inject, Injectable } from '@angular/core';
import { TuiSheetDialogService } from '@taiga-ui/addon-mobile/components/sheet-dialog';
import { TUI_IS_MOBILE, tuiIsString } from '@taiga-ui/cdk';
import { TuiDialogService } from '@taiga-ui/core';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/core";
import * as i2 from "@taiga-ui/addon-mobile/components/sheet-dialog";
/**
 * Same as `TuiDialogService` but automatically switches to `TuiSheetDialogService` on mobile
 */
export class TuiResponsiveDialogService {
    constructor(isMobile, dialogs, sheets) {
        this.isMobile = isMobile;
        this.dialogs = dialogs;
        this.sheets = sheets;
    }
    open(content, options = {}) {
        return this.isMobile
            ? this.sheets.open(content, options)
            : this.dialogs.open(content, Object.assign(Object.assign({}, options), { label: tuiIsString(options.label) ? options.label : '' }));
    }
}
TuiResponsiveDialogService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogService, deps: [{ token: TUI_IS_MOBILE }, { token: TuiDialogService }, { token: TuiSheetDialogService }], target: i0.ɵɵFactoryTarget.Injectable });
TuiResponsiveDialogService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_MOBILE]
                }] }, { type: i1.TuiDialogService, decorators: [{
                    type: Inject,
                    args: [TuiDialogService]
                }] }, { type: i2.TuiSheetDialogService, decorators: [{
                    type: Inject,
                    args: [TuiSheetDialogService]
                }] }]; } });
