// @taiga-ui/addon-mobile/esm2015/directives/responsive-dialog/responsive-dialog.directive.js
import { Directive } from '@angular/core';
import { AbstractTuiDialogDirective, AbstractTuiDialogService } from '@taiga-ui/cdk';
import { TuiResponsiveDialogService } from './responsive-dialog.service';
import * as i0 from "@angular/core";
export class TuiResponsiveDialogDirective extends AbstractTuiDialogDirective {
}
TuiResponsiveDialogDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiResponsiveDialogDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiResponsiveDialogDirective, selector: "ng-template[tuiResponsiveDialog]", inputs: { options: ["tuiResponsiveDialogOptions", "options"], open: ["tuiResponsiveDialog", "open"] }, outputs: { openChange: "tuiResponsiveDialogChange" }, providers: [
        {
            provide: AbstractTuiDialogService,
            useExisting: TuiResponsiveDialogService,
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[tuiResponsiveDialog]',
                    inputs: ['options: tuiResponsiveDialogOptions', 'open: tuiResponsiveDialog'],
                    outputs: ['openChange: tuiResponsiveDialogChange'],
                    providers: [
                        {
                            provide: AbstractTuiDialogService,
                            useExisting: TuiResponsiveDialogService,
                        },
                    ],
                }]
        }] });
