// @taiga-ui/addon-mobile/bundles/taiga-ui-addon-mobile-components-mobile-calendar-dialog.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@tinkoff/ng-polymorpheus'), require('@taiga-ui/addon-mobile/components/mobile-calendar'), require('@taiga-ui/kit'), require('@taiga-ui/addon-mobile/services'), require('@taiga-ui/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-mobile/components/mobile-calendar-dialog', ['exports', '@angular/core', '@taiga-ui/cdk', '@tinkoff/ng-polymorpheus', '@taiga-ui/addon-mobile/components/mobile-calendar', '@taiga-ui/kit', '@taiga-ui/addon-mobile/services', '@taiga-ui/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-mobile"] = global["taiga-ui"]["addon-mobile"] || {}, global["taiga-ui"]["addon-mobile"].components = global["taiga-ui"]["addon-mobile"].components || {}, global["taiga-ui"]["addon-mobile"].components["mobile-calendar-dialog"] = {}), global.ng.core, global.i2, global.ngPolymorpheus, global["taiga-ui"]["addon-mobile"].components["mobile-calendar"], global.i3, global["taiga-ui"]["addon-mobile"].services, global.i4));
})(this, (function (exports, i0, i2, ngPolymorpheus, i1, i3, i5, i4) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i2__namespace = /*#__PURE__*/_interopNamespace(i2);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);
    var i3__namespace = /*#__PURE__*/_interopNamespace(i3);
    var i5__namespace = /*#__PURE__*/_interopNamespace(i5);
    var i4__namespace = /*#__PURE__*/_interopNamespace(i4);

    var TuiMobileCalendarDialogComponent = /** @class */ (function () {
        function TuiMobileCalendarDialogComponent(context) {
            this.context = context;
        }
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "single", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.single) === true;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "multi", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.multi) === true;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "min", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.min) || i2.TUI_FIRST_DAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "max", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.max) || i2.TUI_LAST_DAY;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(TuiMobileCalendarDialogComponent.prototype, "disabledItemHandler", {
            get: function () {
                var _a;
                return ((_a = this.context.data) === null || _a === void 0 ? void 0 : _a.disabledItemHandler) || i2.ALWAYS_FALSE_HANDLER;
            },
            enumerable: false,
            configurable: true
        });
        return TuiMobileCalendarDialogComponent;
    }());
    TuiMobileCalendarDialogComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogComponent, deps: [{ token: ngPolymorpheus.POLYMORPHEUS_CONTEXT }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiMobileCalendarDialogComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileCalendarDialogComponent, selector: "tui-mobile-calendar-dialog", ngImport: i0__namespace, template: "<tui-mobile-calendar\n    [disabledItemHandler]=\"disabledItemHandler\"\n    [max]=\"max\"\n    [min]=\"min\"\n    [multi]=\"multi\"\n    [single]=\"single\"\n    (cancel)=\"context.$implicit.complete()\"\n    (confirm)=\"context.completeWith($event)\"\n></tui-mobile-calendar>\n", styles: [":host{display:block;height:100vh;margin:-1.5rem -1.5rem -2rem}\n"], components: [{ type: i1__namespace.TuiMobileCalendarComponent, selector: "tui-mobile-calendar", inputs: ["single", "multi", "min", "max", "disabledItemHandler", "value"], outputs: ["cancel", "confirm", "valueChange"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-mobile-calendar-dialog',
                        templateUrl: './mobile-calendar-dialog.template.html',
                        styleUrls: ['./mobile-calendar-dialog.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                    }]
            }], ctorParameters: function () {
            return [{ type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [ngPolymorpheus.POLYMORPHEUS_CONTEXT]
                        }] }];
        } });

    var TuiMobileCalendarDropdownComponent = /** @class */ (function () {
        function TuiMobileCalendarDropdownComponent(zone, duration, single, multi, range, dropdown, keyboard) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j;
            this.zone = zone;
            this.duration = duration;
            this.single = single;
            this.multi = multi;
            this.range = range;
            this.dropdown = dropdown;
            this.keyboard = keyboard;
            this.animation = {
                value: '',
                params: {
                    start: '100vh',
                    duration: this.duration,
                },
            };
            this.min = ((_a = this.single) === null || _a === void 0 ? void 0 : _a.min) ||
                ((_b = this.multi) === null || _b === void 0 ? void 0 : _b.min) ||
                ((_c = this.range) === null || _c === void 0 ? void 0 : _c.maxLengthMapper(this.range.computedMin, this.range.value, this.range.maxLength, true)) ||
                i2.TUI_FIRST_DAY;
            this.max = ((_d = this.single) === null || _d === void 0 ? void 0 : _d.max) ||
                ((_e = this.multi) === null || _e === void 0 ? void 0 : _e.max) ||
                ((_f = this.range) === null || _f === void 0 ? void 0 : _f.maxLengthMapper(this.range.computedMax, this.range.value, this.range.maxLength, false)) ||
                i2.TUI_LAST_DAY;
            this.disabledItemHandler = ((_g = this.single) === null || _g === void 0 ? void 0 : _g.disabledItemHandler) ||
                ((_h = this.multi) === null || _h === void 0 ? void 0 : _h.disabledItemHandler) ||
                ((_j = this.range) === null || _j === void 0 ? void 0 : _j.disabledItemHandler) ||
                i2.ALWAYS_FALSE_HANDLER;
            this.keyboard.hide();
        }
        TuiMobileCalendarDropdownComponent.prototype.close = function () {
            this.dropdown.computedHost.focus();
            this.dropdown.updateOpen(false);
            this.keyboard.show();
        };
        TuiMobileCalendarDropdownComponent.prototype.confirm = function (value) {
            var control = this.single || this.multi || this.range;
            if (control) {
                control.value =
                    this.range && value instanceof i2.TuiDay
                        ? new i2.TuiDayRange(value, value)
                        : value;
            }
            this.close();
        };
        return TuiMobileCalendarDropdownComponent;
    }());
    TuiMobileCalendarDropdownComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDropdownComponent, deps: [{ token: i2.TuiActiveZoneDirective }, { token: i4.TUI_ANIMATIONS_DURATION }, { token: i3.TuiInputDateComponent, optional: true }, { token: i3.TuiInputDateMultiComponent, optional: true }, { token: i3.TuiInputDateRangeComponent, optional: true }, { token: i4.TuiHostedDropdownComponent }, { token: i5.TuiKeyboardService }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiMobileCalendarDropdownComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiMobileCalendarDropdownComponent, selector: "tui-mobile-calendar-dropdown", host: { properties: { "@tuiSlideInTop": "this.animation", "@tuiFadeIn": "this.animation" } }, ngImport: i0__namespace, template: "<tui-mobile-calendar\n    tuiActiveZone\n    [disabledItemHandler]=\"disabledItemHandler\"\n    [max]=\"max\"\n    [min]=\"min\"\n    [multi]=\"!!multi\"\n    [single]=\"!!single\"\n    (cancel)=\"close()\"\n    (confirm)=\"confirm($event)\"\n></tui-mobile-calendar>\n", styles: [":host{position:fixed;top:0;left:0;width:100%;height:100%;background:var(--tui-elevation-01);box-shadow:0 10rem var(--tui-elevation-01),0 -90vh 1rem 2rem #000000bf}\n"], components: [{ type: i1__namespace.TuiMobileCalendarComponent, selector: "tui-mobile-calendar", inputs: ["single", "multi", "min", "max", "disabledItemHandler", "value"], outputs: ["cancel", "confirm", "valueChange"] }], directives: [{ type: i2__namespace.TuiActiveZoneDirective, selector: "[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)", inputs: ["tuiActiveZoneParent"], outputs: ["tuiActiveZoneChange"], exportAs: ["tuiActiveZone"] }], animations: [i4.tuiSlideInTop, i4.tuiFadeIn], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDropdownComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-mobile-calendar-dropdown',
                        templateUrl: './mobile-calendar-dropdown.template.html',
                        styleUrls: ['./mobile-calendar-dropdown.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        animations: [i4.tuiSlideInTop, i4.tuiFadeIn],
                    }]
            }], ctorParameters: function () {
            return [{ type: i2__namespace.TuiActiveZoneDirective, decorators: [{
                            type: i0.Inject,
                            args: [i2.TuiActiveZoneDirective]
                        }] }, { type: undefined, decorators: [{
                            type: i0.Inject,
                            args: [i4.TUI_ANIMATIONS_DURATION]
                        }] }, { type: i3__namespace.TuiInputDateComponent, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiInputDateComponent]
                        }] }, { type: i3__namespace.TuiInputDateMultiComponent, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiInputDateMultiComponent]
                        }] }, { type: i3__namespace.TuiInputDateRangeComponent, decorators: [{
                            type: i0.Optional
                        }, {
                            type: i0.Inject,
                            args: [i3.TuiInputDateRangeComponent]
                        }] }, { type: i4__namespace.TuiHostedDropdownComponent, decorators: [{
                            type: i0.Inject,
                            args: [i4.TuiHostedDropdownComponent]
                        }] }, { type: i5__namespace.TuiKeyboardService, decorators: [{
                            type: i0.Inject,
                            args: [i5.TuiKeyboardService]
                        }] }];
        }, propDecorators: { animation: [{
                    type: i0.HostBinding,
                    args: ['@tuiSlideInTop']
                }, {
                    type: i0.HostBinding,
                    args: ['@tuiFadeIn']
                }] } });

    var TuiMobileCalendarDialogModule = /** @class */ (function () {
        function TuiMobileCalendarDialogModule() {
        }
        return TuiMobileCalendarDialogModule;
    }());
    TuiMobileCalendarDialogModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiMobileCalendarDialogModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogModule, declarations: [TuiMobileCalendarDialogComponent, TuiMobileCalendarDropdownComponent], imports: [i1.TuiMobileCalendarModule, i2.TuiActiveZoneModule], exports: [TuiMobileCalendarDialogComponent] });
    TuiMobileCalendarDialogModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogModule, providers: [
            {
                provide: i3.TUI_MOBILE_CALENDAR,
                useValue: TuiMobileCalendarDropdownComponent,
            },
        ], imports: [[i1.TuiMobileCalendarModule, i2.TuiActiveZoneModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiMobileCalendarDialogModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.TuiMobileCalendarModule, i2.TuiActiveZoneModule],
                        declarations: [TuiMobileCalendarDialogComponent, TuiMobileCalendarDropdownComponent],
                        providers: [
                            {
                                provide: i3.TUI_MOBILE_CALENDAR,
                                useValue: TuiMobileCalendarDropdownComponent,
                            },
                        ],
                        exports: [TuiMobileCalendarDialogComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiMobileCalendarDialogComponent = TuiMobileCalendarDialogComponent;
    exports.TuiMobileCalendarDialogModule = TuiMobileCalendarDialogModule;
    exports.TuiMobileCalendarDropdownComponent = TuiMobileCalendarDropdownComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
