// @taiga-ui/addon-mobile/bundles/taiga-ui-addon-mobile-services.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@taiga-ui/cdk'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-mobile/services', ['exports', '@angular/common', '@angular/core', '@taiga-ui/cdk', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-mobile"] = global["taiga-ui"]["addon-mobile"] || {}, global["taiga-ui"]["addon-mobile"].services = {}), global.ng.common, global.ng.core, global.cdk, global.rxjs, global.rxjs.operators));
})(this, (function (exports, common, i0, cdk, rxjs, operators) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var TuiKeyboardService = /** @class */ (function () {
        function TuiKeyboardService(doc) {
            var _this = this;
            this.doc = doc;
            this.sub = rxjs.merge(rxjs.fromEvent(this.doc, 'focusout'), rxjs.fromEvent(this.doc, 'mousedown').pipe(operators.filter(function (e) { return Object.is(e.target, _this.element); }))).subscribe(function () {
                _this.show();
            });
            this.inputMode = '';
        }
        TuiKeyboardService.prototype.ngOnDestroy = function () {
            this.sub.unsubscribe();
            this.show();
        };
        TuiKeyboardService.prototype.toggle = function () {
            if (this.element) {
                this.show();
            }
            else {
                this.hide();
            }
        };
        TuiKeyboardService.prototype.hide = function () {
            var focused = cdk.tuiGetNativeFocused(this.doc);
            if ((focused === null || focused === void 0 ? void 0 : focused.inputMode) === undefined || this.element) {
                return;
            }
            this.element = focused;
            this.inputMode = focused.inputMode;
            focused.inputMode = 'none';
        };
        TuiKeyboardService.prototype.show = function () {
            if (!this.element) {
                return;
            }
            this.element.inputMode = 'none';
            this.element.inputMode = this.inputMode;
            this.element = undefined;
        };
        return TuiKeyboardService;
    }());
    TuiKeyboardService.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeyboardService, deps: [{ token: common.DOCUMENT }], target: i0__namespace.ɵɵFactoryTarget.Injectable });
    TuiKeyboardService.ɵprov = i0__namespace.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeyboardService, providedIn: 'root' });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiKeyboardService, decorators: [{
                type: i0.Injectable,
                args: [{
                        providedIn: 'root',
                    }]
            }], ctorParameters: function () {
            return [{ type: Document, decorators: [{
                            type: i0.Inject,
                            args: [common.DOCUMENT]
                        }] }];
        } });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiKeyboardService = TuiKeyboardService;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
