// @taiga-ui/addon-mobile/bundles/taiga-ui-addon-mobile-components-bottom-sheet.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/addon-mobile/components/bottom-sheet', ['exports', '@angular/core', '@taiga-ui/cdk', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global["taiga-ui"] = global["taiga-ui"] || {}, global["taiga-ui"]["addon-mobile"] = global["taiga-ui"]["addon-mobile"] || {}, global["taiga-ui"]["addon-mobile"].components = global["taiga-ui"]["addon-mobile"].components || {}, global["taiga-ui"]["addon-mobile"].components["bottom-sheet"] = {}), global.ng.core, global.cdk, global.ng.common));
})(this, (function (exports, i0, cdk, i1) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var i1__namespace = /*#__PURE__*/_interopNamespace(i1);

    var OPTIONS = {
        duration: 20,
        easing: 'ease-in',
        fill: 'forwards',
    };
    var TuiBottomSheetComponent = /** @class */ (function () {
        function TuiBottomSheetComponent(el) {
            this.el = el;
            this.elements = cdk.EMPTY_QUERY;
            this.stops = ['1.5rem'];
        }
        TuiBottomSheetComponent.prototype.onScroll = function () {
            var _a, _b;
            var _c = this.el.nativeElement, clientHeight = _c.clientHeight, scrollTop = _c.scrollTop, scrollHeight = _c.scrollHeight;
            var top = ((_a = this.elements.get(0)) === null || _a === void 0 ? void 0 : _a.nativeElement.clientHeight) || 0;
            var max = ((_b = this.content) === null || _b === void 0 ? void 0 : _b.nativeElement.clientHeight) || Infinity;
            var height = Math.min(clientHeight, max);
            var scrolled = Math.min(scrollTop, height - top);
            var transform = "translate3d(0, " + -1 * scrolled + "px, 0)";
            this.el.nativeElement.style.setProperty('--t-height', scrollHeight + "px");
            this.el.nativeElement.animate([{ transform: transform }], OPTIONS);
        };
        return TuiBottomSheetComponent;
    }());
    TuiBottomSheetComponent.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBottomSheetComponent, deps: [{ token: i0.ElementRef }], target: i0__namespace.ɵɵFactoryTarget.Component });
    TuiBottomSheetComponent.ɵcmp = i0__namespace.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBottomSheetComponent, selector: "tui-bottom-sheet", inputs: { stops: "stops" }, host: { listeners: { "scroll.silent": "onScroll()" }, properties: { "style.--t-start": "stops[0]", "style.scroll-snap-type": "stops.length > 1 ? \"y mandatory\" : null" } }, viewQueries: [{ propertyName: "content", first: true, predicate: ["content"], descendants: true }, { propertyName: "elements", predicate: ["stops"], descendants: true }], ngImport: i0__namespace, template: "<div\n    *ngFor=\"let stop of stops\"\n    #stops\n    class=\"t-stop\"\n    [style.top]=\"stop\"\n></div>\n<div\n    #content\n    class=\"t-content\"\n>\n    <ng-content></ng-content>\n</div>\n<div class=\"t-bottom\"></div>\n", styles: [":host{scrollbar-width:none;-ms-overflow-style:none;position:absolute;top:100%;left:0;right:0;display:block;block-size:calc(100% - 1rem);max-inline-size:40rem;margin:calc(-1 * var(--t-start)) auto 0;background:var(--tui-elevation-01);border-top-left-radius:var(--tui-radius-l);border-top-right-radius:var(--tui-radius-l);overflow:auto;box-shadow:0 1rem var(--tui-elevation-01),var(--tui-shadow)}:host::-webkit-scrollbar,:host::-webkit-scrollbar-thumb{display:none}:host:before{content:\"\";position:-webkit-sticky;position:sticky;top:.625rem;left:50%;z-index:1;display:block;inline-size:2rem;block-size:.25rem;-webkit-margin-after:-.25rem;margin-block-end:-.25rem;transform:translate(-50%);background:var(--tui-elevation-01);box-shadow:0 0 0 20rem var(--tui-elevation-01),inset 0 0 0 1rem var(--tui-base-03);border-radius:1rem;-webkit-clip-path:polygon(-100vw -1rem,100vw -1rem,100vw calc(100% + .625rem),-100vw calc(100% + .625rem));clip-path:polygon(-100vw -1rem,100vw -1rem,100vw calc(100% + .625rem),-100vw calc(100% + .625rem))}:host:after{content:\"\";position:absolute;top:calc(100% - var(--t-start));inline-size:1rem;block-size:calc(var(--t-height) - 100% + var(--t-start));scroll-snap-align:start;scroll-snap-stop:always}.t-content{position:-webkit-sticky;position:sticky;top:0;padding:2.25rem 1rem 1.5rem}.t-bottom{block-size:calc(100% - var(--t-start))}.t-stop{position:absolute;block-size:var(--t-start);inline-size:1rem;pointer-events:none;transform:translateY(-100%);scroll-snap-align:start;scroll-snap-stop:always}.t-stop:last-child{scroll-snap-align:none}\n"], directives: [{ type: i1__namespace.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0__namespace.ChangeDetectionStrategy.OnPush });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBottomSheetComponent, decorators: [{
                type: i0.Component,
                args: [{
                        selector: 'tui-bottom-sheet',
                        templateUrl: './bottom-sheet.template.html',
                        styleUrls: ['./bottom-sheet.style.less'],
                        changeDetection: i0.ChangeDetectionStrategy.OnPush,
                        host: {
                            '[style.--t-start]': 'stops[0]',
                            '[style.scroll-snap-type]': 'stops.length > 1 ? "y mandatory" : null',
                            '(scroll.silent)': 'onScroll()',
                        },
                    }]
            }], ctorParameters: function () {
            return [{ type: i0__namespace.ElementRef, decorators: [{
                            type: i0.Inject,
                            args: [i0.ElementRef]
                        }] }];
        }, propDecorators: { elements: [{
                    type: i0.ViewChildren,
                    args: ['stops']
                }], content: [{
                    type: i0.ViewChild,
                    args: ['content']
                }], stops: [{
                    type: i0.Input
                }] } });

    var TuiBottomSheetModule = /** @class */ (function () {
        function TuiBottomSheetModule() {
        }
        return TuiBottomSheetModule;
    }());
    TuiBottomSheetModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBottomSheetModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    TuiBottomSheetModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBottomSheetModule, declarations: [TuiBottomSheetComponent], imports: [i1.CommonModule], exports: [TuiBottomSheetComponent] });
    TuiBottomSheetModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBottomSheetModule, imports: [[i1.CommonModule]] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0__namespace, type: TuiBottomSheetModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        imports: [i1.CommonModule],
                        declarations: [TuiBottomSheetComponent],
                        exports: [TuiBottomSheetComponent],
                    }]
            }] });

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TuiBottomSheetComponent = TuiBottomSheetComponent;
    exports.TuiBottomSheetModule = TuiBottomSheetModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
