// @taiga-ui/addon-mobile/fesm2015/taiga-ui-addon-mobile-directives-responsive-dialog.js
import * as i0 from '@angular/core';
import { Injectable, Inject, Directive, NgModule } from '@angular/core';
import { tuiIsString, TUI_IS_MOBILE, AbstractTuiDialogDirective, AbstractTuiDialogService } from '@taiga-ui/cdk';
import * as i2 from '@taiga-ui/addon-mobile/components/sheet-dialog';
import { TuiSheetDialogService } from '@taiga-ui/addon-mobile/components/sheet-dialog';
import * as i1 from '@taiga-ui/core';
import { TuiDialogService } from '@taiga-ui/core';

/**
 * Same as `TuiDialogService` but automatically switches to `TuiSheetDialogService` on mobile
 */
class TuiResponsiveDialogService {
    constructor(isMobile, dialogs, sheets) {
        this.isMobile = isMobile;
        this.dialogs = dialogs;
        this.sheets = sheets;
    }
    open(content, options = {}) {
        return this.isMobile
            ? this.sheets.open(content, options)
            : this.dialogs.open(content, Object.assign(Object.assign({}, options), { label: tuiIsString(options.label) ? options.label : '' }));
    }
}
TuiResponsiveDialogService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogService, deps: [{ token: TUI_IS_MOBILE }, { token: TuiDialogService }, { token: TuiSheetDialogService }], target: i0.ɵɵFactoryTarget.Injectable });
TuiResponsiveDialogService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [TUI_IS_MOBILE]
                }] }, { type: i1.TuiDialogService, decorators: [{
                    type: Inject,
                    args: [TuiDialogService]
                }] }, { type: i2.TuiSheetDialogService, decorators: [{
                    type: Inject,
                    args: [TuiSheetDialogService]
                }] }]; } });

class TuiResponsiveDialogDirective extends AbstractTuiDialogDirective {
}
TuiResponsiveDialogDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogDirective, deps: null, target: i0.ɵɵFactoryTarget.Directive });
TuiResponsiveDialogDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.2.17", type: TuiResponsiveDialogDirective, selector: "ng-template[tuiResponsiveDialog]", inputs: { options: ["tuiResponsiveDialogOptions", "options"], open: ["tuiResponsiveDialog", "open"] }, outputs: { openChange: "tuiResponsiveDialogChange" }, providers: [
        {
            provide: AbstractTuiDialogService,
            useExisting: TuiResponsiveDialogService,
        },
    ], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ng-template[tuiResponsiveDialog]',
                    inputs: ['options: tuiResponsiveDialogOptions', 'open: tuiResponsiveDialog'],
                    outputs: ['openChange: tuiResponsiveDialogChange'],
                    providers: [
                        {
                            provide: AbstractTuiDialogService,
                            useExisting: TuiResponsiveDialogService,
                        },
                    ],
                }]
        }] });

class TuiResponsiveDialogModule {
}
TuiResponsiveDialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiResponsiveDialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogModule, declarations: [TuiResponsiveDialogDirective], exports: [TuiResponsiveDialogDirective] });
TuiResponsiveDialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiResponsiveDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TuiResponsiveDialogDirective],
                    exports: [TuiResponsiveDialogDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiResponsiveDialogDirective, TuiResponsiveDialogModule, TuiResponsiveDialogService };
