// @taiga-ui/addon-mobile/fesm2015/taiga-ui-addon-mobile-components-bottom-sheet.js
import * as i0 from '@angular/core';
import { ElementRef, Component, ChangeDetectionStrategy, Inject, ViewChildren, ViewChild, Input, NgModule } from '@angular/core';
import { EMPTY_QUERY } from '@taiga-ui/cdk';
import * as i1 from '@angular/common';
import { CommonModule } from '@angular/common';

const OPTIONS = {
    duration: 20,
    easing: 'ease-in',
    fill: 'forwards',
};
class TuiBottomSheetComponent {
    constructor(el) {
        this.el = el;
        this.elements = EMPTY_QUERY;
        this.stops = ['1.5rem'];
    }
    onScroll() {
        var _a, _b;
        const { clientHeight, scrollTop, scrollHeight } = this.el.nativeElement;
        const top = ((_a = this.elements.get(0)) === null || _a === void 0 ? void 0 : _a.nativeElement.clientHeight) || 0;
        const max = ((_b = this.content) === null || _b === void 0 ? void 0 : _b.nativeElement.clientHeight) || Infinity;
        const height = Math.min(clientHeight, max);
        const scrolled = Math.min(scrollTop, height - top);
        const transform = `translate3d(0, ${-1 * scrolled}px, 0)`;
        this.el.nativeElement.style.setProperty('--t-height', `${scrollHeight}px`);
        this.el.nativeElement.animate([{ transform }], OPTIONS);
    }
}
TuiBottomSheetComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetComponent, deps: [{ token: ElementRef }], target: i0.ɵɵFactoryTarget.Component });
TuiBottomSheetComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: TuiBottomSheetComponent, selector: "tui-bottom-sheet", inputs: { stops: "stops" }, host: { listeners: { "scroll.silent": "onScroll()" }, properties: { "style.--t-start": "stops[0]", "style.scroll-snap-type": "stops.length > 1 ? \"y mandatory\" : null" } }, viewQueries: [{ propertyName: "content", first: true, predicate: ["content"], descendants: true }, { propertyName: "elements", predicate: ["stops"], descendants: true }], ngImport: i0, template: "<div\n    *ngFor=\"let stop of stops\"\n    #stops\n    class=\"t-stop\"\n    [style.top]=\"stop\"\n></div>\n<div\n    #content\n    class=\"t-content\"\n>\n    <ng-content></ng-content>\n</div>\n<div class=\"t-bottom\"></div>\n", styles: [":host{scrollbar-width:none;-ms-overflow-style:none;position:absolute;top:100%;left:0;right:0;display:block;block-size:calc(100% - 1rem);max-inline-size:40rem;margin:calc(-1 * var(--t-start)) auto 0;background:var(--tui-elevation-01);border-top-left-radius:var(--tui-radius-l);border-top-right-radius:var(--tui-radius-l);overflow:auto;box-shadow:0 1rem var(--tui-elevation-01),var(--tui-shadow)}:host::-webkit-scrollbar,:host::-webkit-scrollbar-thumb{display:none}:host:before{content:\"\";position:-webkit-sticky;position:sticky;top:.625rem;left:50%;z-index:1;display:block;inline-size:2rem;block-size:.25rem;-webkit-margin-after:-.25rem;margin-block-end:-.25rem;transform:translate(-50%);background:var(--tui-elevation-01);box-shadow:0 0 0 20rem var(--tui-elevation-01),inset 0 0 0 1rem var(--tui-base-03);border-radius:1rem;-webkit-clip-path:polygon(-100vw -1rem,100vw -1rem,100vw calc(100% + .625rem),-100vw calc(100% + .625rem));clip-path:polygon(-100vw -1rem,100vw -1rem,100vw calc(100% + .625rem),-100vw calc(100% + .625rem))}:host:after{content:\"\";position:absolute;top:calc(100% - var(--t-start));inline-size:1rem;block-size:calc(var(--t-height) - 100% + var(--t-start));scroll-snap-align:start;scroll-snap-stop:always}.t-content{position:-webkit-sticky;position:sticky;top:0;padding:2.25rem 1rem 1.5rem}.t-bottom{block-size:calc(100% - var(--t-start))}.t-stop{position:absolute;block-size:var(--t-start);inline-size:1rem;pointer-events:none;transform:translateY(-100%);scroll-snap-align:start;scroll-snap-stop:always}.t-stop:last-child{scroll-snap-align:none}\n"], directives: [{ type: i1.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'tui-bottom-sheet',
                    templateUrl: './bottom-sheet.template.html',
                    styleUrls: ['./bottom-sheet.style.less'],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[style.--t-start]': 'stops[0]',
                        '[style.scroll-snap-type]': 'stops.length > 1 ? "y mandatory" : null',
                        '(scroll.silent)': 'onScroll()',
                    },
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }]; }, propDecorators: { elements: [{
                type: ViewChildren,
                args: ['stops']
            }], content: [{
                type: ViewChild,
                args: ['content']
            }], stops: [{
                type: Input
            }] } });

class TuiBottomSheetModule {
}
TuiBottomSheetModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBottomSheetModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetModule, declarations: [TuiBottomSheetComponent], imports: [CommonModule], exports: [TuiBottomSheetComponent] });
TuiBottomSheetModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiBottomSheetComponent],
                    exports: [TuiBottomSheetComponent],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TuiBottomSheetComponent, TuiBottomSheetModule };
