// @taiga-ui/addon-mobile/esm2015/services/keyboard.service.js
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import { tuiGetNativeFocused } from '@taiga-ui/cdk';
import { fromEvent, merge } from 'rxjs';
import { filter } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class TuiKeyboardService {
    constructor(doc) {
        this.doc = doc;
        this.sub = merge(fromEvent(this.doc, 'focusout'), fromEvent(this.doc, 'mousedown').pipe(filter(e => Object.is(e.target, this.element)))).subscribe(() => {
            this.show();
        });
        this.inputMode = '';
    }
    ngOnDestroy() {
        this.sub.unsubscribe();
        this.show();
    }
    toggle() {
        if (this.element) {
            this.show();
        }
        else {
            this.hide();
        }
    }
    hide() {
        const focused = tuiGetNativeFocused(this.doc);
        if ((focused === null || focused === void 0 ? void 0 : focused.inputMode) === undefined || this.element) {
            return;
        }
        this.element = focused;
        this.inputMode = focused.inputMode;
        focused.inputMode = 'none';
    }
    show() {
        if (!this.element) {
            return;
        }
        this.element.inputMode = 'none';
        this.element.inputMode = this.inputMode;
        this.element = undefined;
    }
}
TuiKeyboardService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeyboardService, deps: [{ token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Injectable });
TuiKeyboardService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeyboardService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiKeyboardService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; } });
