// @taiga-ui/addon-mobile/esm2015/components/bottom-sheet/bottom-sheet.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiBottomSheetComponent } from './bottom-sheet.component';
import * as i0 from "@angular/core";
export class TuiBottomSheetModule {
}
TuiBottomSheetModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TuiBottomSheetModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetModule, declarations: [TuiBottomSheetComponent], imports: [CommonModule], exports: [TuiBottomSheetComponent] });
TuiBottomSheetModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: TuiBottomSheetModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [TuiBottomSheetComponent],
                    exports: [TuiBottomSheetComponent],
                }]
        }] });
